### PR TITLE
Feature/integrate events failure

### PR DESCRIPTION
**Pull request by Jakub-Vacek**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](SECURITY.md) is correct.
- [ ] I have read the **CONTRIBUTION_GUIDE** document.
- [ ] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/client/failure/event-adapter.test.ts [hunk 1]
```diff
@@ -1,11 +1,18 @@
 import { MapDocumentNode, ProfileDocumentNode } from '@superfaceai/ast';
+import {
+  ProfileConfiguration,
+  Provider,
+  ProviderConfiguration,
+  UseCase,
+} from '@superfaceai/one-sdk';
+import { SuperfaceClientBase } from '@superfaceai/one-sdk/dist/client/public/client';
```
Comment (lukas-valenta) [inline src/client/failure/event-adapter.test.ts:8]: these are not the imports you're looking for :grin:

Comment (Jakub-Vacek) [reply src/client/failure/event-adapter.test.ts:8]: Yep, this file is a mess. I am rewriting it now 🙂

File: src/client/client.ts [hunk 2]
```diff
@@ -27,17 +31,14 @@ export abstract class SuperfaceClientBase {
 
     if (SUPER_CACHE[superCacheKey] === undefined) {
       SUPER_CACHE[superCacheKey] = SuperJson.loadSync(superCacheKey).unwrap();
+    } else {
+      //TODO: better way of cummunicating this to user.
+      console.warn('Multiple SuperfaceClient bad');
```
Comment (TheEdward162) [inline src/client/client.ts:36]: 🤣

File: src/client/client.ts [hunk 3]
```diff
@@ -133,8 +134,128 @@ export abstract class SuperfaceClientBase {
...
+      { providersOfUsecase: Record<string, FailurePolicy> }
+    > = {};
+    let policy: FailurePolicy;
+    for (const [profile, profileSettings] of Object.entries(
+      this.superJson.normalized.profiles
+    )) {
+      //Set failoverPolicy
+      const priority = profileSettings.priority;
+      if (!priority.every(p => this.superJson.normalized.providers[p])) {
+        throw new SDKExecutionError(
```
Comment (TheEdward162) [inline src/client/client.ts:176]: Duplicate code :/

File: src/client/usecase.ts [hunk 4]
```diff
@@ -70,20 +76,20 @@ export class UseCase extends UseCaseBase {
     const boundProfileProvider = await this.bind(options);
 
     // TOOD: rewrap the errors for public consumption?
-    return boundProfileProvider.perform<TInput, TOutput>(this.name, input);
+    return await boundProfileProvider.perform<TInput, TOutput>(this.name, input);
```
Comment (TheEdward162) [inline src/client/usecase.ts:79]: I'm afraid to ask but.. what breaks without this await?

Comment (Jakub-Vacek) [reply src/client/usecase.ts:79]: Nope, I was just trying to localize the concurrency bug :)

File: src/client/failure/policies.ts [hunk 5]
```diff
@@ -12,6 +12,71 @@ import {
   SuccessResolution,
 } from './resolution';
 
+export class Router {
+  constructor(
+    private readonly providersOfUseCase: Record<string, FailurePolicy>,
+    private readonly priority: string[],
+    private currentProvider?: string
```
Comment (TheEdward162) [inline src/client/failure/policies.ts:19]: Why is this optional?

File: src/client/failure/policies.ts [hunk 6]
```diff
@@ -12,6 +12,71 @@ import {
...
+  public afterSuccess(info: ExecutionSuccess): SuccessResolution {
+    if (!this.currentProvider) {
+      throw 'Property currentProvider is not set in Router instance';
+    }
+
+    return this.providersOfUseCase[this.currentProvider].afterSuccess(info);
+  }
+
+  //TODO: Should we handle provider reset here?
+  public reset(): void {}
```
Comment (TheEdward162) [inline src/client/failure/policies.ts:77]: Imo this should reset itself to the first provider and reset all inner policies

File: src/client/failure/policies.ts [hunk 7]
```diff
@@ -12,6 +12,71 @@ import {
   SuccessResolution,
 } from './resolution';
 
+export class Router {
```
Comment (TheEdward162) [inline src/client/failure/policies.ts:15]: Can we please name this something more descriptive? Like `FailurePolicyRouter`.

File: src/client/failure/policies.ts [hunk 8]
```diff
@@ -12,6 +12,71 @@ import {
...
+    // TODO: check if exists
+    this.currentProvider = provider;
+  }
+
+  public beforeExecution(info: ExecutionInfo): ExecutionResolution {
+    if (!this.currentProvider) {
+      throw 'Property currentProvider is not set in Router instance';
+    }
+    const innerResolution =
+      this.providersOfUseCase[this.currentProvider].beforeExecution(info);
```
Comment (TheEdward162) [inline src/client/failure/policies.ts:36]: At some point we should also attempt going back to previous providers. I think an iteration of the return values of `beforeExecution` here should do the trick.

Comment (Jakub-Vacek) [reply src/client/failure/policies.ts:61]: @TheEdward162 I added something very similar and also test for it.

**Pull request merged**